module/zfs/dbuf.c

```diff
@@ -1358,14 +1415,22 @@ dbuf_read(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 		spa_t *spa = dn->dn_objset->os_spa;
 
 		/*
-		 * If the arc buf is compressed or encrypted, we need to
-		 * untransform it to read the data. This could happen during
-		 * the "zfs receive" of a stream which is deduplicated and
-		 * either raw or compressed. We do not need to do this if the
-		 * caller wants raw encrypted data.
+		 * Ensure that this block's dnode has been decrypted if
+		 * the caller has requested decrypted data.
 		 */
-		if (db->db_buf != NULL && (flags & DB_RF_NO_DECRYPT) == 0 &&
+		err = dbuf_read_verify_dnode_crypt(db, flags);
+
+		/*
+		 * If the arc buf is compressed or encrypted and the caller
+		 * requested uncompressed data, we need to untransform it
+		 * before returning. We also attempt to  untransform any
+		 * unauthenticated blocks that we may have cached if we now
+		 * have the key available.
```

Comment:
Is this really just an attempt, if we have the key available?  If the key is unavailable, or the authentication fails, then that's fine?  I would think that would cause arc_untransform() to fail, and this read will fail.

Reply:
The arc authentication here works as follows:

1. This function calls `arc_untransform()` if it sees this is an unauthenticated buffer
2. `arc_untransform()` attempts to authenticate the buffer
2a. if we succeed, we simply unset the "unathenticated" bit in the arc header and return 0
2b. If we fail because the key isn't loaded, we still return 0, but leave the "unathenticated" bit set in the arc header
2c. if there is any other failure, we set return EIO, set the IO_ERROR bit in the arc header and create zpool events

Reply:
I see.  Since we always call arc_untransform() on unauthenticated blocks (i.e. "untransform them"), could you reword the comment to say something like: `We also call arc_untransform() on any unauthenticated blocks, which will verify their MAC if the key is now available.`

Reply:
Fixed.